source/opt/scalar_replacement_pass.cpp

```diff
@@ -0,0 +1,702 @@
+  for (auto var : replacements) {
+    // Create a load of each replacement variable.
+    ir::Instruction* type = GetStorageType(var);
+    uint32_t loadId = TakeNextId();
+    std::unique_ptr<ir::Instruction> newLoad(
+        new ir::Instruction(context(), SpvOpLoad, type->result_id(), loadId,
+                            std::initializer_list<ir::Operand>{
+                                {SPV_OPERAND_TYPE_ID, {var->result_id()}}}));
+    // Copy memory access attributes.
+    for (uint32_t i = 1; i < load->NumInOperands(); ++i) {
+      ir::Operand copy(load->GetInOperand(i));
+      newLoad->AddOperand(std::move(copy));
+    }
+    where = where.InsertBefore(std::move(newLoad));
+    get_def_use_mgr()->AnalyzeInstDefUse(&*where);
+    context()->set_instr_block(&*where, block);
+    loads.push_back(&*where);
+  }
+
+  // Construct a new composite.
+  uint32_t compositeId = TakeNextId();
+  where = load;
+  std::unique_ptr<ir::Instruction> compositeConstruct(new ir::Instruction(
+      context(), SpvOpCompositeConstruct, load->type_id(), compositeId, {}));
+  for (auto l : loads) {
+    ir::Operand op(SPV_OPERAND_TYPE_ID,
+                   std::initializer_list<uint32_t>{l->result_id()});
+    compositeConstruct->AddOperand(std::move(op));
+  }
+  where = where.InsertBefore(std::move(compositeConstruct));
+  get_def_use_mgr()->AnalyzeInstDefUse(&*where);
+  context()->set_instr_block(&*where, block);
+  context()->ReplaceAllUsesWith(load->result_id(), compositeId);
+}
+
+void ScalarReplacementPass::ReplaceWholeStore(
+    ir::Instruction* store, const std::vector<ir::Instruction*>& replacements) {
+  // Replaces a store to the whole composite with a series of extract and stores
+  // to each element.
+  uint32_t storeInput = store->GetSingleWordInOperand(1u);
+  ir::BasicBlock* block = context()->get_instr_block(store);
+  ir::BasicBlock::iterator where(store);
+  uint32_t elementIndex = 0;
+  for (auto var : replacements) {
+    // Create the extract.
+    ir::Instruction* type = GetStorageType(var);
+    uint32_t extractId = TakeNextId();
+    std::unique_ptr<ir::Instruction> extract(new ir::Instruction(
+        context(), SpvOpCompositeExtract, type->result_id(), extractId,
+        std::initializer_list<ir::Operand>{
+            {SPV_OPERAND_TYPE_ID, {storeInput}},
+            {SPV_OPERAND_TYPE_LITERAL_INTEGER, {elementIndex++}}}));
+    auto iter = where.InsertBefore(std::move(extract));
+    get_def_use_mgr()->AnalyzeInstDefUse(&*iter);
+    context()->set_instr_block(&*iter, block);
+
+    // Create the store.
+    std::unique_ptr<ir::Instruction> newStore(
+        new ir::Instruction(context(), SpvOpStore, 0, 0,
+                            std::initializer_list<ir::Operand>{
+                                {SPV_OPERAND_TYPE_ID, {var->result_id()}},
+                                {SPV_OPERAND_TYPE_ID, {extractId}}}));
+    // Copy memory access attributes.
+    for (uint32_t i = 2; i < store->NumInOperands(); ++i) {
```

Comment:
Why skipping the first two?  This needs a comment.

Reply:
Ok